modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java

```diff
@@ -0,0 +1,204 @@
+package org.motechproject.metrics.service.impl;
+
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.RatioGauge;
+import org.motechproject.metrics.api.Counter;
+import org.motechproject.metrics.api.Gauge;
+import org.motechproject.metrics.api.Histogram;
+import org.motechproject.metrics.api.Meter;
+import org.motechproject.metrics.api.Metric;
+import org.motechproject.metrics.api.Timer;
+import org.motechproject.metrics.config.MetricsConfigFacade;
+import org.motechproject.metrics.model.CounterAdapter;
+import org.motechproject.metrics.model.Enablable;
+import org.motechproject.metrics.model.GaugeAdapter;
+import org.motechproject.metrics.model.HistogramAdapter;
+import org.motechproject.metrics.model.MeterAdapter;
+import org.motechproject.metrics.model.TimerAdapter;
+import org.motechproject.metrics.service.MetricRegistryService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+@Service("metricRegistryService")
+public class MetricRegistryServiceImpl implements MetricRegistryService {
+    private final MetricRegistry metricRegistry;
+    private final MetricsConfigFacade metricsConfigFacade;
+
+    private Map<String, Metric> metrics;
+
+    @Autowired
+    public MetricRegistryServiceImpl(MetricRegistry metricRegistry, MetricsConfigFacade metricsConfigFacade) {
+        this.metricRegistry = metricRegistry;
+        this.metricsConfigFacade = metricsConfigFacade;
+        metrics = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void setEnabled(boolean enabled) {
+        for (Metric metric: metrics.values()) {
+            if (metric instanceof Enablable && ((Enablable) metric).isEnabled() != enabled) {
+                ((Enablable) metric).setEnabled(enabled);
+            }
+        }
+    }
+
+    @Override
+    public Counter counter(final String name) {
+        return getOrAdd(name, MetricBuilder.COUNTERS);
+    }
+
+    @Override
+    public Histogram histogram(final String name) {
+        return getOrAdd(name, MetricBuilder.HISTOGRAMS);
+    }
+
+    @Override
+    public Meter meter(final String name) {
+        return getOrAdd(name, MetricBuilder.METERS);
+    }
+
+    @Override
+    public Timer timer(final String name) {
+        return getOrAdd(name, MetricBuilder.TIMERS);
+    }
+
+    @Override
+    public <T> Gauge<T> registerGauge(final String name, final Gauge<T> gauge) {
+        com.codahale.metrics.Gauge<T> theGauge;
+        try {
+            theGauge = metricRegistry.register(name, new com.codahale.metrics.Gauge<T>() {
+                @Override
+                public T getValue() {
+                    return gauge.getValue();
+                }
+            });
+        } catch (IllegalArgumentException ex) {
+            throw ex;
```

Comment:
Doesn't make sense to catch the exception only to throw the same exception again. Did you mean to wrap it into custom exception?